File: avalon-modules/avalon-dataset/avalon-dataset-service/src/main/java/io/github/lishangbu/avalon/dataset/service/impl/StatServiceImpl.java
package io.github.lishangbu.avalon.dataset.service.impl;

import io.github.lishangbu.avalon.dataset.entity.Stat;
import io.github.lishangbu.avalon.dataset.entity.Stat_;
import io.github.lishangbu.avalon.dataset.repository.StatRepository;
import io.github.lishangbu.avalon.dataset.service.StatService;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.ExampleMatcher;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

/// 能力服务实现。
@Service
@RequiredArgsConstructor
public class StatServiceImpl implements StatService {
    private final StatRepository statRepository;

    @Override
    public Page<Stat> getPageByCondition(Stat stat, Pageable pageable) {
        return statRepository.findAll(
                Example.of(
                        stat,
                        ExampleMatcher.matching()
                                .withIgnoreNullValues()
                                .withMatcher(
                                        Stat_.NAME,
                                        ExampleMatcher.GenericPropertyMatchers.contains())
                                .withMatcher(
                                        Stat_.INTERNAL_NAME,
                                        ExampleMatcher.GenericPropertyMatchers.contains())),
                pageable);
    }

    @Override
    public List<Stat> listByCondition(Stat stat) {
        ExampleMatcher matcher =
                ExampleMatcher.matching()
                        .withIgnoreNullValues()
                        .withMatcher(
                                Stat_.NAME,
                                ExampleMatcher.GenericPropertyMatchers.contains())
                        .withMatcher(
                                Stat_.INTERNAL_NAME,
                                ExampleMatcher.GenericPropertyMatchers.contains());
        return statRepository.findAll(Example.of(stat, matcher));
    }

    @Override
    public Stat save(Stat stat) {
        return statRepository.save(stat);
    }

    @Override
    public Stat update(Stat stat) {
        return statRepository.save(stat);
    }

    @Override
    public void removeById(Long id) {
        statRepository.deleteById(id);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-service/src/main/kotlin/io/github/lishangbu/avalon/dataset/service/impl/StatServiceImpl.kt
package io.github.lishangbu.avalon.dataset.service.impl

import io.github.lishangbu.avalon.dataset.entity.Stat
import io.github.lishangbu.avalon.dataset.entity.dto.SaveStatInput
import io.github.lishangbu.avalon.dataset.entity.dto.StatSpecification
import io.github.lishangbu.avalon.dataset.entity.dto.StatView
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateStatInput
import io.github.lishangbu.avalon.dataset.repository.StatRepository
import io.github.lishangbu.avalon.dataset.service.StatService
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.springframework.stereotype.Service

/** 能力值服务实现*/
@Service
class StatServiceImpl(
    /** 能力值仓储*/
    private val statRepository: StatRepository,
) : StatService {
    /** 根据条件查询能力值列表*/
    override fun listByCondition(specification: StatSpecification): List<StatView> = statRepository.listViews(specification)

    /** 保存能力值*/
    override fun save(command: SaveStatInput): StatView = statRepository.save(command.toEntity(), SaveMode.INSERT_ONLY).let(::reloadView)

    /** 更新能力值*/
    override fun update(command: UpdateStatInput): StatView {
        requireStatEditable(command.id.toLong(), "能力值已设为只读，禁止修改")
        return statRepository.save(command.toEntity(), SaveMode.UPSERT).let(::reloadView)
    }

    /** 按 ID 删除能力值*/
    override fun removeById(id: Long) {
        requireStatEditable(id, "能力值已设为只读，禁止删除")
        statRepository.deleteById(id)
    }

    private fun reloadView(stat: Stat): StatView = requireNotNull(statRepository.loadViewById(stat.id)) { "未找到 ID=${stat.id} 对应的能力值" }

    private fun requireStatEditable(
        id: Long,
        errorMessage: String,
    ) {
        val stat = requireNotNull(statRepository.findNullable(id)) { "未找到 ID=$id 对应的能力值" }
        if (stat.readonly) {
            throw IllegalStateException(errorMessage)
        }
    }
}
